### PR TITLE
Serialization & Types improvements

### LINKED ISSUE #133: When listing a Component's model, show the function signature of a value
If the value of a property of a Vue data model is a function. Vue Devtools currently just shows `"Function"` as the value. It would be neat if it showed the function signature instead (eg. `"function someFunction(arg)"`).

You can get the signature by calling `.toString()` on the function and getting everything up to the closing parenthesis:

```
if (typeof value === "function") {
    const valueStr = value.toString()
    return valueStr.slice(0, valueStr.indexOf(")") + 1)
}
```

### LINKED ISSUE #349: show structure Set
code like this
```javascript
data() {
   setKeys: new Set(),
}
```
vue-devtools will be show this
image
better cast **new Set()** to structure **Array** in show result

### LINKED ISSUE #499: Cannot read property 'type' of null
OS : Ubuntu 16.04
Chrome : 63.0.3239.84
vue : 2.5.13
vuex : 3.0.1
vue-devtools: 4.0.0

No content is appeared at right side, when I click on a component to inspect it. 
After I click on a component to inspect it, there are also a lot of errors in my console.

Please look at these two images below.

image

image

### LINKED ISSUE #507: v4.0.1 - "Uncaught TypeError: Cannot read property '_route' of undefined" Everywhere
On chrome devtools -> Vue, if I click on any component (on the component tree on the left side) this error appears on console:

`vue-router.esm.js?880d:557 Uncaught TypeError: Cannot read property '_route' of undefined
    at Vue$3.get (vue-router.esm.js?880d:557)
    at backend.js:1
    at x (backend.js:1)
    at t.a (backend.js:1)
    at Object.y (backend.js:1)
    at e (backend.js:1)
    at e (backend.js:1)
    at e (backend.js:1)
    at e (backend.js:1)
    at Object.t.stringifyStrict (backend.js:1)`

updated the the project packages, including vue-router, tried on an old version of the project that was working before, but with the v4.0.1 it gives the same error. Installed an old version of vue-devtools (3.1.7) and everything works well as before.

when events or other actions are triggered when using the website the console fills with errors:

On firefox it gives this error:
TypeError: this._routerRoot is undefined

### LINKED ISSUE #528: Uncaught RangeError: Maximum call stack size exceeded at Array.toString (native)
### Version
4.0.0

### Browser and OS info
Chrome 63.0.3239.132 (Official Build) (64-bit) / Windows 10

### Steps to reproduce
- Go to this [url](https://2wk5wvxrj.codesandbox.io/) to view the sample project on CodeSandbox.io
- Right-click the page and select "Inspect"
- In the Developer Tools window, click Vue menu to display Vue-Devtools
- Click on `<app>` on the left pane, the right pane should not cause any error
- Click on `<Vuetable>` node or its child component `<VuetableRowHeader>`, and it dumps error out in the console pane

### What is expected?
Vue-Devtools should display info about the selected component.

Before 4.0.0, this never happens.

### What is actually happening?
The following error shows up in console pane

```
backend.js:1 Uncaught RangeError: Maximum call stack size exceeded
    at Array.toString (native)
    at D (chrome-extension://nhdogjmejiglipccpnnnanhbledajbpd/build/backend.js:1:20128)
...
```
 

---
The old version of Vue-Devtools (before 4.0) was working fine without any error showing up.

### PR DESCRIPTION
**Pull request by Akryum**
- CustomValue API changes
  - new `tooltip` option
  - fields order retained from `value`
- Special format (using CustomValue API) for:
  - Vuex Store
  - VueRouter
  - Component definition
  - Function (Fix #133)
  - Set & Map (Fix #349, Original PR #490)
- Serialization is more fault tolerant (Fix #499 Fix #507)
- Cache Custom values to prevent stack overflow (circular references, Fix #528)
- Switch the order of 'data' and 'props' sections in component inspector to better follow [the style guide](https://vuejs.org/v2/style-guide/#Component-instance-options-order-recommended)
- Performance improvements (related to #515) 

capture d ecran 2018-01-16 a 12 27 33
capture d ecran 2018-01-16 a 12 28 23
capture d ecran 2018-01-16 a 12 29 20

### CODE REVIEW

**Comment by Akryum** [comment]
@evan @posva @michalsnik I fixed #528 with a new `EncodeCache` class that should be used in front of any serialization involving CustomValue API. I also changed the order of data and props sections to better follow [the style guide](https://vuejs.org/v2/style-guide/#Component-instance-options-order-recommended). WDYT?

**Comment by Akryum** [comment]
capture d ecran du 2018-01-19 12 45 30

**Comment by Akryum** [comment]
I added some optimization from this branch: https://github.com/Akryum/vue-devtools/tree/component-inspect-optimizations ([changes](https://github.com/Akryum/vue-devtools/compare/a68ee61e07acf44d4ea7cb3b20a0e2493e734824...Akryum:component-inspect-optimizations))

x10 faster when selecting a component (tested with the `NativeTypes` component after clicking the `Create large array` button).

**Comment by Akryum** [comment]
@yyx990803 I also made a PR https://github.com/yyx990803/circular-json-es6/pull/5 with the changes I made to CircularJSON.

**Comment by Akryum** [comment]
I don't think we can handle WeakMap since they are not iterable.

File: src/util.js [hunk 1]
```diff
@@ -52,29 +58,184 @@ export const SPECIAL_TOKENS = {
...
+    }
+  }
+}
+
+export function getCustomFunctionDetails (func) {
+  const args = func.toString().match(/\(.*\)/)[0]
+  return {
+    _custom: {
+      type: 'function',
+      display: `<span>${func.name}</span>${args}`
```
Comment (posva) [inline src/util.js]: I think adding the word _function_ somewhere could make this more explicit. What about something like _ƒ o(a, b, c)_, similar to the console?

**Comment by posva** [comment]
Couldn't test open in editor yet, but everything else looks fine 🙂

**Comment by Akryum** [comment]
capture d ecran du 2018-01-20 23 33 31

File: src/util.js [hunk 2]
```diff
@@ -52,29 +58,184 @@ export const SPECIAL_TOKENS = {
...
-    return `[native Date ${val.toString()}]`
-  } else if (val && val._isVue) {
-    return getCustomInstanceDetails(val)
+  }
+  return sanitize(val)
+}
+
+export function getCustomMapDetails (val) {
+  const list = []
+  val.forEach(
```
Comment (michalsnik) [inline src/util.js]: You could use `map` instead:
```
const list = val.map((value, key) => ({ key, value}))
```

File: src/util.js [hunk 3]
```diff
@@ -52,29 +58,184 @@ export const SPECIAL_TOKENS = {
...
+        abstract: true
+      }
+    }
+  }
+}
+
+export function reviveMap (val) {
+  const result = new Map()
+  const list = val._custom.value
+  for (let i = 0; i < list.length; i++) {
```
Comment (michalsnik) [inline src/util.js]: Perhaps you can use `forEach` instead? 
```
list.forEach(({ key, value}) =>
  result.set(key, reviver(null, value)
)
```

File: src/util.js [hunk 4]
```diff
@@ -52,29 +58,184 @@ export const SPECIAL_TOKENS = {
...
+      value: list,
+      readOnly: true
+    }
+  }
+}
+
+export function reviveSet (val) {
+  const result = new Set()
+  const list = val._custom.value
+  for (let i = 0; i < list.length; i++) {
```
Comment (michalsnik) [inline src/util.js]: Same here

File: src/devtools/env.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  ctrl: isMac ? '&#8984;' : 'Ctrl',
+  shift: 'Shift',
+  alt: isMac ? '&#8997;' : 'Alt'
+}
+
+Object.defineProperties(Vue.prototype, {
+  '$isChrome': { get: () => isChrome },
+  '$isWindows': { get: () => isWindows },
+  '$isMac': { get: () => isMac },
+  '$isLinus': { get: () => isLinux },
```
Comment (michalsnik) [inline src/devtools/env.js]: Shouldn't it be `$isLinux`?

File: shells/dev/target/NativeTypes.vue [hunk 6]
```diff
@@ -11,43 +11,116 @@
...
   },
+
   data () {
     return {
       localDate: new Date(),
-      testComponent: null
+      reg: /abc/gi,
+      testComponent: null,
+      hello: function foo (a, b, c) {},
+      hey: function empty () {},
```
Comment (michalsnik) [inline shells/dev/target/NativeTypes.vue]: Can you also add anonymous function here and make sure it works good as well?
`hi: function (a, b) {}` should be displayed probably as  `f (a, b)`, also what about arrow functions? Or `hi() {}`?

File: src/util.js (hunk 2 above)
Comment (Akryum) [reply src/util.js]: `Map` doesn't have any `map` method.

File: src/util.js (hunk 3 above)
Comment (Akryum) [reply src/util.js]: Simple loop without anonymous function call is faster imho. Also made a test case: https://jsperf.com/map-loop-vs-foreach (for what it's worth :upside_down_face:)

**Comment by Akryum** [comment]
@michalsnik 
capture d ecran du 2018-01-22 11 24 10

**Comment by Akryum** [comment]
@michalsnik 
capture d ecran du 2018-01-22 11 25 10

**Pull request merged**